File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/array/randomized/P710_RandomPickWithBlacklist.java
package com.dsa.leetcode.labuladong.array.randomized;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Random;
import java.util.Set;

public class P710_RandomPickWithBlacklist {

    class Solution {
        private Map<Integer, Integer> b2w;
        private int bound;
        private Random random;

        public Solution(int n, int[] blacklist) {
            random = new Random();
            bound = n - blacklist.length;

            Set<Integer> black = new HashSet<>();
            for (int b : blacklist) {
                if (b >= bound) {
                    black.add(b);
                }
            }

            Integer w = bound;
            b2w = new HashMap<>();
            for (int b : blacklist) {
                if (b < bound) {
                    while (black.contains(w)) {
                        w++;
                    }
                    b2w.put(b, w);
                    w++;
                }
            }
        }

        public int pick() {
            int x = random.nextInt(bound);
            return b2w.getOrDefault(x, x);
        }
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/array/randomized/P710_RandomPickWithBlacklist.py
# author: tobebetter9527
# since: 2024/2/21 8:49
import random
from typing import List


class Solution:

    def __init__(self, n: int, blacklist: List[int]):
        bound = n - len(blacklist)
        self.bound = bound

        black = set()
        for b in blacklist:
            if b >= bound:
                black.add(b)

        w = bound
        b2w = {}
        for b in blacklist:
            if b < bound:
                while w in black:
                    w += 1
                b2w[b] = w
                w += 1
        self.b2w = b2w

    def pick(self) -> int:
        x = random.randrange(self.bound)
        return self.b2w.get(x, x)
